EIPS/eip-1571.md

```diff
@@ -0,0 +1,552 @@
+
+Using proper error codes pools may properly inform miners of the condition of their requests. Error codes **MUST** honor this scheme:
+
+- Error codes 2xx : request accepted and processed but some additional info in the `error` member may give hint
+- Error codes 3xx : server could not process the request due to a lack of authorization by the client
+- Error codes 4xx : server could not process the request due to syntactic problems (method not found, missing params, wrong data types ecc.) or passed `param` values are not valid.
+- Error codes 5xx : server could not process the request due to internal errors
+
+### Notifications
+A notification message has the very same representation of a `request` with the only difference the `id` member **MUST NOT** be present. This means the issuer is not interested nor expects any reponse to this message. It's up to the receiver to take actions accordingly. For instance the receiver **MAY** decide to execute the method, or, in case of errors or methods not allowed, drop the connection thus closing the session.
+
+#### Error member
+As seen above a `response` **MAY** contain an `error` member. When present this member **MUST** be an Object with:
+- mandatory member `code` : a Number which identifies the error occurred
+- mandatory member `message` : a short human readable sentence describing the error occurred
+- optional member `data` : a Structured or Primitive value that contains additional informations about the error. The value of this member is defined by the Server (e.g. detailed error information, nested errors etc.).
+
+## Protocol Flow
+- Client starts session by opening a TCP socket to the server
+- Client advertises and request protocol compatibility
+- Server confirms compatibility and declares ready
+- Client starts/resumes a session
+- Client sends request for authorization for each of it's workers
+- Server replies back with responses for each authorization
+- Server sends `mining.set` for constant values to be adopted for following mining jobs
+- Server sends `mining.notify` with minimal job info
+- Client mines on job
+- Client sends `mining.submit` if any solution found for the job
+- Server replies whether solution is accepted
+- Server optionally sends `mining.set` for constant values to be adopted for following mining jobs (if something changed)
+- Server sends `mining.notify` with minimal job info
+- ... (continue)
+- Eventually either party closes session and TCP connection
+
+### Session Handling - Hello
+~~One of the worst annoyances until now is that server, at the very moment of socket connection, does not provide any useful information about the stratum flavour implemented. This means the client has to start a conversation by iteratively trying to connect via different protocol flavours. This proposal amends the situation making mandatory for the server to advertise itself to the client. 
```

Comment:
Does this strikethrough paragraph need to be here?

Reply:
No it doesn't. Can be removed